### PR TITLE
Data type Validation for Custom Fields

### PR DESCRIPTION
**Pull request by cgalarza**

### CODE REVIEW

File: app/models/term.rb [hunk 1]
```diff
@@ -46,21 +46,36 @@ def set_custom_field(field, value)
...
-  def already_exists?
-    errors.added?(:uri, :uniquness) || errors.added?(:uri_hash, :uniqueness)
-  end
-
   private
 
     def validate_custom_fields
-      custom_fields.each do |k, _|
+      custom_fields.each do |k, v|
         # TODO: Validate data_type.
```
Comment (elohanlon) [inline app/models/term.rb]: Leftover comment?

**Review by elohanlon — APPROVED** [review]

**Comment by cgalarza** [comment]
I forgot to add type casting in a `:before_save` hook. Need to go add that in!

File: app/models/term.rb (hunk 1 above)
Comment (cgalarza) [reply app/models/term.rb]: yup! thx

**Comment by cgalarza** [comment]
Adding the before_save hook to cast values after they have been validation. 

@elohanlon I am wondering if i should throw errors if the values can't be cast to their correct values because they aren't strings or because they are invalid in some other way.

**Comment by elohanlon** [comment]
@cgalarza That's a good idea about throwing errors.  Even though it's unlikely that validation won't be run before a save, I like how that kind of protective code would stop bad data from entering our datastore.  👍

**Comment by cgalarza** [comment]
@elohanlon Raising errors when invalid custom field values are still present when saving. Added a few more tests. Let me know if its ready to merge.

File: app/models/term.rb [hunk 2]
```diff
@@ -46,21 +46,58 @@ def set_custom_field(field, value)
...
+        raise "custom_field #{k} is not a valid custom field" unless vocabulary.custom_fields.keys.include?(k)
+
+        case vocabulary.custom_fields[k][:data_type]
+        when 'string'
+          raise "custom_field #{k} must be a string" unless v.is_a?(String)
+        when 'integer'
+          if valid_integer?(v)
+            custom_fields[k] = v.to_i if v.is_a?(String)
+          else
+            raise "custom_field #{k} must be an integer"
```
Comment (elohanlon) [inline app/models/term.rb]:
```suggestion
            raise "custom_field #{k} must be an integer. Could not cast String \"#{v}\" to an integer."
```
How do you feel about this proposed addition?

File: app/models/term.rb [hunk 3]
```diff
@@ -46,21 +46,58 @@ def set_custom_field(field, value)
...
+          end
+        when 'boolean'
+          if valid_boolean?(v)
+            custom_fields[k] = (v == 'true') ? true : false if v.is_a?(String)
+          else
+            raise "custom_field #{k} must be a boolean"
```
Comment (elohanlon) [inline app/models/term.rb]:
```suggestion
            raise "custom_field #{k} must be a boolean. Could not cast String \"#{v}\" to a boolean."
```
How do you feel about this proposed addition?

Comment (cgalarza) [reply app/models/term.rb]: The additions both look good. I'll add them in.

**Comment by cgalarza** [comment]
@elohanlon Changes made. Let me know if I can merge

**Comment by elohanlon** [comment]
@cgalarza Merge away!  Thanks for the change.

**Review by elohanlon — APPROVED** [review]

**Pull request merged**